### PR TITLE
[MWPW-137148] [PERFORMANCE] Load IMS from the same domain

### PR DESCRIPTION
**Pull request by mokimo**
## Description
- Selfhost IMS to avoid DNS lookups wasting precious time for visitors. 
- Check for a new version of imslib.min.js every 24h and automatically create a PR on updates [(see example)](https://github.com/mokimo/milo/pull/9) based on the last modified date of the resource

## Related Issue
Resolves: [MWPW-137148](https://jira.corp.adobe.com/browse/MWPW-137148)

## Testing instructions
1. Signing in should still function
2. Signing out should still work
3. Regression tests on all consumers for IMS related functionality

## Runnings this action locally
Development can be cumbersome, for this reason you can also run this on your local environment using node.

1. create a `.env` file on the root of your repo with `GH_TOKEN, REPO_OWNER, REPO_NAME, LOCAL_RUN=true` & if you want to test the workflow for a script other than IMS, you can set a few more configs.
```
GH_TOKEN= // https://docs.github.com/en/authentication/keeping-your-account-and-data-secure/managing-your-personal-access-tokens
REPO_OWNER=mokimo
REPO_NAME=milo
LOCAL_RUN=true
LOCAL_RUN_ORIGIN=mokimo
LOCAL_RUN_BRANCH=test_runs
LOCAL_RUN_SCRIPT=https://assets.adobedtm.com/d4d114c60e50/a0e989131fd5/launch-a27b33fc2dc0-development.min.js
LOCAL_RUN_SCRIPT_PATH=./libs/deps/launch-a27b33fc2dc0-development.min.js
LOCAL_RUN_TITLTE=[AUTOMATED-PR] Update launch prod dependency
``` 
2. run `node --env-file=.env .github/workflows/update-script.js`

## Test URLs
**Acrobat:**
- Before: https://www.stage.adobe.com/acrobat/online/sign-pdf.html?martech=off
- After: https://www.stage.adobe.com/acrobat/online/sign-pdf.html?martech=off&milolibs=mwpw-137148-same-domain-ims--milo--mokimo

**BACOM:**
- Before: https://business.stage.adobe.com/fr/customer-success-stories.html?martech=off
- After: https://business.stage.adobe.com/fr/customer-success-stories.html?martech=off&milolibs=mwpw-137148-same-domain-ims--milo--mokimo

**CC:**
- Before: https://main--cc--adobecom.hlx.live/?martech=off
- After: https://main--cc--adobecom.hlx.live/?martech=off&milolibs=mwpw-137148-same-domain-ims--milo--mokimo

**Homepage:**
- Before: https://main--homepage--adobecom.hlx.page/homepage/index-loggedout?martech=off
- After: https://main--homepage--adobecom.hlx.page/homepage/index-loggedout?martech=off&milolibs=mwpw-137148-same-domain-ims--milo--mokimo

**Blog:**
- Before: https://main--blog--adobecom.hlx.page/
- After: https://main--blog--adobecom.hlx.page/?milolibs=mwpw-137148-same-domain-ims--milo--mokimo

**Milo:**
- Before: https://main--milo--mokimo.hlx.live/drafts/ramuntea/gnav-refactor?martech=off
- After: https://mwpw-137148-same-domain-ims--milo--mokimo.hlx.live/drafts/ramuntea/gnav-refactor?martech=off

### CODE REVIEW

**Review by JasonHowellSlavin — APPROVED** [review]
Awesome! Looks good. Do we need to update any tests for checking that IMS is coming from the right place. 

This saves a lot of load time on slow networks. Thank you for tackling this!

File: .github/workflows/update-ims.js [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+
+        res.on('end', () => {
+          resolve({
+            data,
+            headers: res.headers,
+          });
+        });
+      })
+      .on('error', (err) => {
+        reject(err);
```
Comment (overmyheadandbody) [inline .github/workflows/update-ims.js:27]: I'm guessing this is an actual `Error`, not a string, right?

File: .github/workflows/update-ims.js [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+- Before: https://main--cc--adobecom.hlx.live/?martech=off
+- After: https://main--cc--adobecom.hlx.live/?martech=off&milolibs=update-imslib--milo--adobecom
+
+**Homepage:**
+- Before: https://main--homepage--adobecom.hlx.page/homepage/index-loggedout?martech=off
+- After: https://main--homepage--adobecom.hlx.page/homepage/index-loggedout?martech=off&milolibs=update-imslib--milo--adobecom
+
+**Blog:**
+- Before: https://main--blog--adobecom.hlx.page/
+- After: https://main--blog--adobecom.hlx.page/?milolibs=update-imslib--milo--adobecom
```
Comment (overmyheadandbody) [inline .github/workflows/update-ims.js:102]: Negligible nit: should these also have `?martech=off`?

File: .github/workflows/update-ims.js [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+  const diffHours = Math.ceil((now - lastModified) / 1000 / 60 / 60);
+  console.log(
+    `imslib.min.js last modified: ${lastModified}. Last modified ${diffHours} hours ago <= 24H, means new PR created.`
+  );
+
+  if (diffHours <= 24 || local) {
+    console.log('Diff detected in the last 24h, creating PR.');
+    const header = `// Last modified ${lastModified.toISOString()}\n// Build at ${new Date().toISOString()}\n`;
+
+    if (local) return console.log('Local mode - skipped git commands.');
```
Comment (overmyheadandbody) [inline .github/workflows/update-ims.js:44]: Could this not be the first declaration to skip the previous console log and header declaration?

File: .github/workflows/update-ims.js [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+    execSync('git add libs/deps/imslib.min.js');
+    execSync('git commit -m "Update imslib.min.js"');
+    execSync('git push --force origin update-imslib');
+
+    await github.rest.pulls.create({
+      owner: context.repo.owner,
+      repo: context.repo.repo,
+      title: 'Update imslib.min.js',
+      head: 'update-imslib',
+      base: 'stage',
```
Comment (overmyheadandbody) [inline .github/workflows/update-ims.js:70]: Should we consider adding any labels and/or reviewers?

**Review by overmyheadandbody — APPROVED** [review]
Amazing job on streamlining this! This might be useful for other dependencies as well. We should wait to see it in action and figure out any quirks and then maybe open up a discussion documenting the approach for other to use.

File: .github/workflows/update-ims.js (hunk 4 above)
Comment (mokimo) [reply .github/workflows/update-ims.js:70]: I'm on this one currently, it definitely should have labels & reviewers!

File: .github/workflows/update-ims.js (hunk 1 above)
Comment (mokimo) [reply .github/workflows/update-ims.js:27]: Yup it's an actual `Error`

**Comment by mokimo** [comment]
As discussed with @chrischrischris we also want to self-host https://github.com/adobecom/milo/blob/stage/libs/martech/martech.js#L102 for more performance gains. I'll also add this as part of this PR.

File: .github/workflows/localWorkflowConfigs.js [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+
+const getLocalConfigs = () => {
+  if (!owner || !repo || !auth) {
+    throw new Error(`Create a .env file on the root of the project with credentials.
+Then run: node --env-file=.env .github/workflows/update-ims.js`);
+  }
+
+  const { Octokit } = require('@octokit/rest');
+  return {
+    github: { rest: new Octokit({ auth: process.env.GH_TOKEN }) },
```
Comment (overmyheadandbody) [inline .github/workflows/localWorkflowConfigs.js:15]: I'm guessing this can just be `auth`, since you declare it on L5?

**Review by overmyheadandbody — APPROVED** [review]
❤️

**Review by robert-bogos — APPROVED** [review]

**Review by SilviuLCF — APPROVED** [review]
Ready to STAGE

Testing details 
https://jira.corp.adobe.com/browse/MWPW-137148

File: .github/workflows/update-script.js [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+  // When testing locally, u likely do not want to kill your dev branch
+  if (!localExecution) {
+    execSync('git config --global user.name "GitHub Action"');
+    execSync('git config --global user.email "action@github.com"');
+    execSync('git fetch');
+    execSync('git checkout stage');
+    execSyncSafe(`git branch -D ${branch}`);
+    execSync(`git checkout -b ${branch}`);
+  }
+  console.log('writing script to file', scriptPath);
```
Comment (narcis-radu) [inline .github/workflows/update-script.js:88]: Should `imslib.min.js` be updated when running on local? I'm thinking the source of truth should be the upstream project to make sure we're consistent.

**Review by narcis-radu — COMMENTED** [review]
Great work on this, I'm super excited to see the results.

File: .github/workflows/update-script.js [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+  title: process.env.LOCAL_RUN_TITLTE || '[AUTOMATED-PR] Update imslib.min.js dependency',
+  path: process.env.LOCAL_RUN_SCRIPT || 'https://auth.services.adobe.com/imslib/imslib.min.js',
+  scriptPath: process.env.LOCAL_RUN_SCRIPT_PATH || './libs/deps/imslib.min.js',
+  origin: process.env.LOCAL_RUN_ORIGIN || 'origin',
+};
+
+const getPrDescription = ({ branch, scriptPath }) => `## Description
+Update ${scriptPath} to the latest version
+## Related Issue
+Resolves: NO TICKET - AUTOMATED CREATED PR.
```
Comment (narcis-radu) [inline .github/workflows/update-script.js:18]: Can we link to a wiki page or [discussion](https://github.com/orgs/adobecom/discussions/1461)? It should be easier for folks to understand the details.

**Pull request merged**